clayer/pysertype.c

```diff
@@ -1063,76 +1071,124 @@ static inline uint32_t check_number_of_samples (long long n)
   return (uint32_t)n;
 }
 
-static bool readtake_pre (long long N, uint32_t *Nu32, dds_sample_info_t **info, ddspy_sample_container_t **container, ddspy_sample_container_t ***rcontainer)
+static PyObject *readtake_post (int32_t sts, collector_state_t *state)
 {
-  if (!(*Nu32 = check_number_of_samples (N)))
-    return false;
-
-  *info = dds_alloc (sizeof (**info) * *Nu32);
-  *container = dds_alloc (sizeof (**container) * *Nu32);
-  *rcontainer = dds_alloc (sizeof (**rcontainer) * *Nu32);
+  if (sts < 0)
```

Comment:
If some samples have been collected before the error occurred, this will leak. An error during the collection step is a bit of a problem: some states changes occur in the process and it can't roll it back. The options I see are:

* free the memory, return an error and accept that some samples are lost forever (or marked as read but never actually seen by the application)
* return the collected samples and pretend nothing went wrong, on the (reasonable) assumption that the application will try again for the remainder
* return the collected samples while (somehow) recording an error occurred, and returning an error the next time round

I wouldn't go for the third one, it is terribly confusing to get a seemingly correct result from a failing operation, and then an error code much later when there is no apparent cause.

The errors that can occur in practice are running out of memory or failing to deserialize the data. The first is exceedingly unlikely, the second can also be done by just dropping only the failing sample.

So here I would do:
```suggestion
  if (sts < 0 && state->count == 0)
```

Reply:
good catch, i agree with the suggestion to return what we have already successfully collected, the user will happily collect the missing in the next round.